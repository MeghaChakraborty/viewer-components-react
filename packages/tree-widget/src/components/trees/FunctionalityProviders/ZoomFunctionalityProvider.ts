/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/

import { NodeKey } from "@bentley/presentation-common";
import { TreeModelNode } from "@bentley/ui-components";
import { IModelConnection, IModelApp } from "@bentley/imodeljs-frontend";
import { Id64 } from "@bentley/bentleyjs-core";
import { TreeNodeFunctionalityProvider } from "./TreeNodeFunctionalityProvider";

export class ZoomFunctionalityProvider extends TreeNodeFunctionalityProvider {

  public async performAction(node: TreeModelNode) {
    await this.zoomSelected(node);
  }

  private static async getElementProps(
    iModel: IModelConnection,
    elemIds: string[] | string
  ) {
    const arr = [...Id64.toIdSet(elemIds)];
    const limit = 100;
    const promises = [];
    while (arr.length > 0) {
      promises.push(iModel.elements.getProps(arr.splice(0, limit)));
    }
    const arrays = await Promise.all(promises);
    return arrays.reduce((a, b) => a.concat(b), []);
  }

  public async zoomSelected (node: TreeModelNode) {
    const elementKey = this._treeDataProvider.getNodeKey(node.item)
    if (NodeKey.isInstancesNodeKey(elementKey)){
      const elemProps = await ZoomFunctionalityProvider.getElementProps(
        IModelApp.viewManager.selectedView!.iModel,
        elementKey.instanceKeys[0].id
      );
      await IModelApp.viewManager.selectedView!.zoomToElementProps(elemProps);
    }
  }
}
