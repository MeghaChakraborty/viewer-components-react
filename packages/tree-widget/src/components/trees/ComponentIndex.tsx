/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/

import * as React from "react";
import { Ruleset } from "@bentley/presentation-common";
import { ControlledTreeWrapper, populateMapWithCommonMenuItems } from "./TreeWithRuleset";
import { FunctionIconInfo, TreeNodeFunctionIconInfoMapper } from "./FunctionalityProviders";
import { IPresentationTreeDataProvider, PresentationTreeDataProvider } from "@bentley/presentation-components";
import { IModelConnection } from "@bentley/imodeljs-frontend";
import componentIndex from "../rulesets/ComponentIndex.json";
import { TreeWidget as BuildingUIComponents } from "../../TreeWidget";
import { GenericOptionItemHandler, OptionItemHandler } from "./OptionItemHandlers";
import { LoadableRuleSetComponent } from "./LoadableRuleSetComponent";

export interface ComponentIndexProps {
  iModel: IModelConnection;
  displayGuids: boolean;
  setIsDisplayGuids: (displayGuids: boolean) => void;
  enableVisibility?: boolean;
}
const COMPONENT_INDEX_NAME = "ComponentIndex";

export const ComponentIndex: React.FC<ComponentIndexProps> = (props: ComponentIndexProps) => {
  const treeName = "ComponentIndexTree";
  const dataProvider = React.useMemo(() => {
    const provider = new PresentationTreeDataProvider({
      imodel: props.iModel,
      ruleset: componentIndex.id,
    });
    provider.pagingSize = 20; // paging size is now needed for the controlled tree.
    return provider;
  }, [props.iModel],
  );

  const { functionIconMapper, optionItems, displayGuidHandler } = React.useMemo(() => {
    const functionIconMapper = new TreeNodeFunctionIconInfoMapper(dataProvider);
    const optionItems: OptionItemHandler[] = [];

    populateContextMenuItems(treeName, functionIconMapper, dataProvider);

    const displayGuidHandler = new GenericOptionItemHandler("Show Guids", BuildingUIComponents.translate("contextMenu.showGuids"), "icon-label", () => { return props.displayGuids; }, props.setIsDisplayGuids);
    optionItems.push(displayGuidHandler);
    return { functionIconMapper, optionItems, displayGuidHandler };
  }, [dataProvider]);

  displayGuidHandler._getItemState = () => props.displayGuids;

  const componentIndexTree = React.useMemo(() => (<ControlledTreeWrapper
    iModel={props.iModel}
    loadedRuleset={componentIndex as Ruleset}
    dataProvider={dataProvider}
    treeName={COMPONENT_INDEX_NAME} treeNodeIconMapper={functionIconMapper} optionItems={optionItems}
    searchTools={false}
    displayGuids={props.displayGuids}
    setIsDisplayGuids={props.setIsDisplayGuids}
    enableVisibility={props.enableVisibility ? props.enableVisibility : false}
  />), [props.iModel, props.displayGuids, props.setIsDisplayGuids]);

  return (
    <LoadableRuleSetComponent ruleSet={componentIndex as Ruleset} >
      {componentIndexTree}
    </LoadableRuleSetComponent>
  );
}

function populateContextMenuItems(treeName: string, mapper: TreeNodeFunctionIconInfoMapper, dataProvider: IPresentationTreeDataProvider) {
  populateMapWithCommonMenuItems(treeName, mapper, dataProvider, componentIndex.id);
}
