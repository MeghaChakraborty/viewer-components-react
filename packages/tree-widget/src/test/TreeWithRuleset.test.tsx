/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/


import * as React from "react";
import { Provider } from "react-redux";
import { mount, shallow } from "enzyme";
import { IModelApp, NoRenderApp, IModelConnection, SelectionSet } from "@bentley/imodeljs-frontend";
import { BeEvent } from "@bentley/bentleyjs-core";
import { ControlledTreeWrapper } from "../components/trees/TreeWithRuleset";
import Rules from "../components/rulesets/SpatialBreakdown.json";
import { TestUtils, TreeWithRuleSetController, setupDataProvider } from "./Utils";
import { Presentation } from "@bentley/presentation-frontend";
import * as moq from "typemoq";
import { Ruleset, HierarchyUpdateInfo } from "@bentley/presentation-common";
import { PresentationManager, RulesetManager, RulesetVariablesManager } from "@bentley/presentation-frontend";
import { TreeNodeFunctionIconInfoMapper } from "../components/trees/FunctionalityProviders";
import { assert } from "chai";
// import { TreeNodeWrapper } from "../components/trees/NodeRenderers/FunctionalTreeNodeRenderer";
// import { PropertyValueRendererManager, TreeActions, TreeModel, TreeModelNode } from "@bentley/ui-components";
// import sinon from "sinon";

describe("TreeWithRuleset Component tests.", () => {

  const connection = moq.Mock.ofType<IModelConnection>();
  const controller = moq.Mock.ofType<TreeWithRuleSetController>();
  const shutdownIModelApp = () => {
    if (IModelApp.initialized)
      IModelApp.shutdown();
  };
  before(async () => {
    // This is required by our I18n module (specifically the i18next package).
    // (global as any).XMLHttpRequest = require("xmlhttprequest").XMLHttpRequest; // tslint:disable-line:no-var-requires
    shutdownIModelApp();
    NoRenderApp.startup();
    Presentation.terminate();
    Presentation.initialize();

    const selectionSet = moq.Mock.ofType<SelectionSet>();
    selectionSet.setup((x) => x.elements).returns(() => new Set([]));
    connection.setup((x) => x.selectionSet).returns(() => selectionSet.object);
    await TestUtils.initializeUiFramework(connection.object);
    IModelApp.i18n.registerNamespace("TreeWidget");
    controller.setup((x) => x.createDataProvider()).returns(() => setupDataProvider(connection.object));

    function mockStorage() {
      var storage: any = {};
      return {
        setItem: function (key: any, value: any) {
          storage[key] = value || '';
        },
        getItem: function (key: any) {
          return storage[key];
        },
        removeItem: function (key: any) {
          delete storage[key];
        },
        get length() {
          return Object.keys(storage).length;
        },
        key: function (i: any) {
          var keys = Object.keys(storage);
          return keys[i] || null;
        }
      };
    }
    (global as any).sessionStorage = mockStorage();
  });


  const rulesetManagerMock = moq.Mock.ofType<RulesetManager>();
  const rulesetVariablesManagerMock = moq.Mock.ofType<RulesetVariablesManager>();
  const presentationManagerMock = moq.Mock.ofType<PresentationManager>();

  let onHierarchyUpdateEvent: BeEvent<(args: { ruleset: Ruleset, updateInfo: HierarchyUpdateInfo }) => void>;
  beforeEach(async () => {
    onHierarchyUpdateEvent = new BeEvent();
    rulesetVariablesManagerMock.reset();
    presentationManagerMock.reset();
    presentationManagerMock.setup((x) => x.onIModelHierarchyChanged).returns(() => onHierarchyUpdateEvent);
    presentationManagerMock.setup((x) => x.rulesets()).returns(() => rulesetManagerMock.object);
    presentationManagerMock.setup((x) => x.vars(moq.It.isAny())).returns(() => rulesetVariablesManagerMock.object);
    rulesetVariablesManagerMock.setup((x) => x.onVariableChanged).returns(() => new BeEvent());
    Presentation.setPresentationManager(presentationManagerMock.object);
    await Presentation.initialize();
  });

  afterEach(() => {
    Presentation.terminate();
  });

  after(() => {
    (global as any).sessionStorage = undefined;
    TestUtils.terminateUiFramework();
    Presentation.terminate();
    shutdownIModelApp();
  });



  it("TreeWithRuleset renders correctly", () => {
    const dataProvider = setupDataProvider(connection.object);
    const functionIconMapper = new TreeNodeFunctionIconInfoMapper(dataProvider);
    const wrapper = (
      <Provider store={TestUtils.store}>
        <ControlledTreeWrapper
          dataProvider={dataProvider}
          setIsDisplayGuids={(_display: boolean) => { }}
          iModel={connection.object}
          displayGuids={true}
          loadedRuleset={Rules as Ruleset} treeNodeIconMapper={functionIconMapper}
          treeName="TreeWithRulesetTest" optionItems={[]} searchTools={true} enableVisibility={false}
        />
      </Provider>
    );
    shallow(wrapper).should.matchSnapshot();
  });

  it("TreeWithRuleset renders correctly with visibility", () => {
    const dataProvider = setupDataProvider(connection.object);
    const functionIconMapper = new TreeNodeFunctionIconInfoMapper(dataProvider);
    const wrapper = (
      <Provider store={TestUtils.store}>
        <ControlledTreeWrapper
          dataProvider={dataProvider}
          setIsDisplayGuids={(_display: boolean) => { }}
          iModel={connection.object}
          displayGuids={true}
          loadedRuleset={Rules as Ruleset} treeNodeIconMapper={functionIconMapper}
          treeName="TreeWithRulesetTest" optionItems={[]} searchTools={true} enableVisibility={true}
        />
      </Provider>
    );
    shallow(wrapper).should.matchSnapshot();
  });

  it("TreeWithRuleset renders correctly with children", () => {
    const dataProvider = setupDataProvider(connection.object);
    const functionIconMapper = new TreeNodeFunctionIconInfoMapper(dataProvider);
    const wrapper = (
      <Provider store={TestUtils.store}>
        <ControlledTreeWrapper
          dataProvider={setupDataProvider(connection.object)}
          setIsDisplayGuids={(_display: boolean) => { }}
          iModel={connection.object}
          displayGuids={true}
          loadedRuleset={Rules as Ruleset} treeNodeIconMapper={functionIconMapper}
          treeName="TreeWithRulesetTest" optionItems={[]} searchTools={true}
          enableVisibility={false}
        />
      </Provider>
    );
    const mountedWrapper = mount(
      wrapper,
    );
    assert(mountedWrapper.exists(".custom-tree-content"), "should contain the custom-tree-content div");
    assert(mountedWrapper.exists(".custom-tree-toolbar"), "should contain the custom-tree-toolbar div");
    assert(mountedWrapper.exists(".custom-tree-container"), "should contain the custom-tree-container div");
    assert(mountedWrapper.exists(".custom-tree-toolbar"), "should contain the custom-tree-toolbar div");
    assert(mountedWrapper.exists("SearchBar"), "should contain the SearchBar component");
    assert(mountedWrapper.exists("SearchBox"), "should contain the SearchBox component");
    assert(mountedWrapper.exists("ControlledTree"), "should contain the ControlledTree component");
  });

});
