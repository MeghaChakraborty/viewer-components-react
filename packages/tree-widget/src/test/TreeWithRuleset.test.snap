// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeWithRuleset Component tests. TreeWithRuleset renders correctly 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "@@observable": [Function],
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "@@observable": [Function],
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <Component
    dataProvider={
      Object {
        "dispose": [Function],
        "getFilteredNodePaths": [Function],
        "getNodeKey": [Function],
        "getNodes": [Function],
        "getNodesCount": [Function],
        "imodel": null,
        "loadHierarchy": [Function],
        "rulesetId": "ui-framework/RulesList",
      }
    }
    displayGuids={true}
    enableVisibility={false}
    iModel={null}
    loadedRuleset={
      Object {
        "$schema": "../../node_modules/@bentley/presentation-common/Ruleset.schema.json",
        "id": "ui-framework/SpatialBreakdown",
        "rules": Array [
          Object {
            "ruleType": "RootNodes",
            "specifications": Array [
              Object {
                "arePolymorphic": true,
                "classes": Array [
                  Object {
                    "classNames": Array [
                      "CompositeElement",
                    ],
                    "schemaName": "SpatialComposition",
                  },
                ],
                "groupByClass": false,
                "groupByLabel": false,
                "instanceFilter": "NOT model.IsPrivate AND this.ComposingElement = NULL",
                "relatedInstances": Array [
                  Object {
                    "alias": "model",
                    "class": Object {
                      "className": "Model",
                      "schemaName": "BisCore",
                    },
                    "isRequired": true,
                    "relationship": Object {
                      "className": "ModelContainsElements",
                      "schemaName": "BisCore",
                    },
                    "requiredDirection": "Backward",
                  },
                ],
                "specType": "InstanceNodesOfSpecificClasses",
              },
            ],
          },
          Object {
            "condition": "ParentNode.IsOfClass(\\"CompositeElement\\", \\"SpatialComposition\\")",
            "ruleType": "ChildNodes",
            "specifications": Array [
              Object {
                "doNotSort": true,
                "groupByClass": false,
                "groupByLabel": false,
                "relatedClasses": Object {
                  "classNames": Array [
                    "DistributionSystem",
                  ],
                  "schemaName": "DistributionSystems",
                },
                "relationships": Object {
                  "classNames": Array [
                    "PhysicalSystemServicesSpatialElements",
                  ],
                  "schemaName": "BisCore",
                },
                "requiredDirection": "Backward",
                "specType": "RelatedInstanceNodes",
              },
            ],
          },
          Object {
            "condition": "ParentNode.IsOfClass(\\"DistributionSystem\\", \\"DistributionSystems\\")",
            "ruleType": "ChildNodes",
            "specifications": Array [
              Object {
                "doNotSort": true,
                "groupByClass": false,
                "groupByLabel": false,
                "relatedClasses": Object {
                  "classNames": Array [
                    "SpatialElement",
                  ],
                  "schemaName": "Biscore",
                },
                "relationships": Object {
                  "classNames": Array [
                    "DistributionSystemGroupsDistributionElements",
                  ],
                  "schemaName": "DistributionSystems",
                },
                "requiredDirection": "Forward",
                "specType": "RelatedInstanceNodes",
              },
            ],
          },
          Object {
            "condition": "ParentNode.IsOfClass(\\"CompositeElement\\", \\"SpatialComposition\\")",
            "ruleType": "ChildNodes",
            "specifications": Array [
              Object {
                "doNotSort": true,
                "groupByClass": false,
                "groupByLabel": false,
                "relatedClasses": Object {
                  "classNames": Array [
                    "CompositeElement",
                  ],
                  "schemaName": "SpatialComposition",
                },
                "relationships": Object {
                  "classNames": Array [
                    "CompositeComposesSubComposites",
                  ],
                  "schemaName": "SpatialComposition",
                },
                "requiredDirection": "Forward",
                "specType": "RelatedInstanceNodes",
              },
            ],
          },
          Object {
            "condition": "ParentNode.IsOfClass(\\"CompositeElement\\", \\"SpatialComposition\\")",
            "ruleType": "ChildNodes",
            "specifications": Array [
              Object {
                "groupByClass": true,
                "groupByLabel": false,
                "relatedClasses": Object {
                  "classNames": Array [
                    "SpatialElement",
                  ],
                  "schemaName": "BisCore",
                },
                "relationships": Object {
                  "classNames": Array [
                    "CompositeOverlapsSpatialElements",
                  ],
                  "schemaName": "SpatialComposition",
                },
                "requiredDirection": "Forward",
                "specType": "RelatedInstanceNodes",
              },
            ],
          },
          Object {
            "condition": "ParentNode.IsOfClass(\\"CompositeElement\\", \\"SpatialComposition\\")",
            "ruleType": "ChildNodes",
            "specifications": Array [
              Object {
                "groupByClass": true,
                "groupByLabel": false,
                "relatedClasses": Object {
                  "classNames": Array [
                    "SpatialElement",
                  ],
                  "schemaName": "BisCore",
                },
                "relationships": Object {
                  "classNames": Array [
                    "SpatialOrganizerReferencesSpatialElements",
                  ],
                  "schemaName": "SpatialComposition",
                },
                "requiredDirection": "Forward",
                "specType": "RelatedInstanceNodes",
              },
            ],
          },
          Object {
            "condition": "ParentNode.IsOfClass(\\"CompositeElement\\", \\"SpatialComposition\\")",
            "ruleType": "ChildNodes",
            "specifications": Array [
              Object {
                "groupByClass": true,
                "groupByLabel": false,
                "relatedClasses": Object {
                  "classNames": Array [
                    "SpatialElement",
                  ],
                  "schemaName": "BisCore",
                },
                "relationships": Object {
                  "classNames": Array [
                    "SpatialOrganizerHoldsSpatialElements",
                  ],
                  "schemaName": "SpatialComposition",
                },
                "requiredDirection": "Forward",
                "specType": "RelatedInstanceNodes",
              },
            ],
          },
          Object {
            "condition": "ContentDisplayType = \\"RelatedElementsRequest\\"",
            "ruleType": "Content",
            "specifications": Array [
              Object {
                "relationshipPaths": Array [
                  Array [
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "CompositeComposesSubComposites",
                        "schemaName": "SpatialComposition",
                      },
                    },
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "CompositeOverlapsSpatialElements",
                        "schemaName": "SpatialComposition",
                      },
                    },
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "ElementOwnsChildElements",
                        "schemaName": "BisCore",
                      },
                    },
                  ],
                ],
                "specType": "ContentRelatedInstances",
              },
              Object {
                "relationshipPaths": Array [
                  Array [
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "CompositeComposesSubComposites",
                        "schemaName": "SpatialComposition",
                      },
                      "targetClass": Object {
                        "className": "ISpatialOrganizer",
                        "schemaName": "SpatialComposition",
                      },
                    },
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "SpatialOrganizerHoldsSpatialElements",
                        "schemaName": "SpatialComposition",
                      },
                    },
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "ElementOwnsChildElements",
                        "schemaName": "BisCore",
                      },
                    },
                  ],
                ],
                "specType": "ContentRelatedInstances",
              },
              Object {
                "relationshipPaths": Array [
                  Array [
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "CompositeComposesSubComposites",
                        "schemaName": "SpatialComposition",
                      },
                    },
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "SpatialOrganizerReferencesSpatialElements",
                        "schemaName": "SpatialComposition",
                      },
                    },
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "ElementOwnsChildElements",
                        "schemaName": "BisCore",
                      },
                    },
                  ],
                ],
                "specType": "ContentRelatedInstances",
              },
              Object {
                "relationshipPaths": Array [
                  Array [
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "DistributionSystemGroupsDistributionElements",
                        "schemaName": "DistributionSystems",
                      },
                    },
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "ElementOwnsChildElements",
                        "schemaName": "BisCore",
                      },
                    },
                  ],
                ],
                "specType": "ContentRelatedInstances",
              },
            ],
          },
        ],
      }
    }
    optionItems={Array []}
    searchTools={true}
    setIsDisplayGuids={[Function]}
    treeName="TreeWithRulesetTest"
    treeNodeIconMapper={
      TreeNodeFunctionIconInfoMapper {
        "_TreeNodeClassFunctionIconInfoMap": Map {},
        "_functionIconInfos": Array [],
        "_globalFunctionIconInfos": Array [],
        "_groupNodeFunctionIconInfos": Array [],
        "_treeDataProvider": Object {
          "dispose": [Function],
          "getFilteredNodePaths": [Function],
          "getNodeKey": [Function],
          "getNodes": [Function],
          "getNodesCount": [Function],
          "imodel": null,
          "loadHierarchy": [Function],
          "rulesetId": "ui-framework/RulesList",
        },
      }
    }
  />
</ContextProvider>
`;

exports[`TreeWithRuleset Component tests. TreeWithRuleset renders correctly with visibility 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "@@observable": [Function],
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "@@observable": [Function],
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <Component
    dataProvider={
      Object {
        "dispose": [Function],
        "getFilteredNodePaths": [Function],
        "getNodeKey": [Function],
        "getNodes": [Function],
        "getNodesCount": [Function],
        "imodel": null,
        "loadHierarchy": [Function],
        "rulesetId": "ui-framework/RulesList",
      }
    }
    displayGuids={true}
    enableVisibility={true}
    iModel={null}
    loadedRuleset={
      Object {
        "$schema": "../../node_modules/@bentley/presentation-common/Ruleset.schema.json",
        "id": "ui-framework/SpatialBreakdown",
        "rules": Array [
          Object {
            "ruleType": "RootNodes",
            "specifications": Array [
              Object {
                "arePolymorphic": true,
                "classes": Array [
                  Object {
                    "classNames": Array [
                      "CompositeElement",
                    ],
                    "schemaName": "SpatialComposition",
                  },
                ],
                "groupByClass": false,
                "groupByLabel": false,
                "instanceFilter": "NOT model.IsPrivate AND this.ComposingElement = NULL",
                "relatedInstances": Array [
                  Object {
                    "alias": "model",
                    "class": Object {
                      "className": "Model",
                      "schemaName": "BisCore",
                    },
                    "isRequired": true,
                    "relationship": Object {
                      "className": "ModelContainsElements",
                      "schemaName": "BisCore",
                    },
                    "requiredDirection": "Backward",
                  },
                ],
                "specType": "InstanceNodesOfSpecificClasses",
              },
            ],
          },
          Object {
            "condition": "ParentNode.IsOfClass(\\"CompositeElement\\", \\"SpatialComposition\\")",
            "ruleType": "ChildNodes",
            "specifications": Array [
              Object {
                "doNotSort": true,
                "groupByClass": false,
                "groupByLabel": false,
                "relatedClasses": Object {
                  "classNames": Array [
                    "DistributionSystem",
                  ],
                  "schemaName": "DistributionSystems",
                },
                "relationships": Object {
                  "classNames": Array [
                    "PhysicalSystemServicesSpatialElements",
                  ],
                  "schemaName": "BisCore",
                },
                "requiredDirection": "Backward",
                "specType": "RelatedInstanceNodes",
              },
            ],
          },
          Object {
            "condition": "ParentNode.IsOfClass(\\"DistributionSystem\\", \\"DistributionSystems\\")",
            "ruleType": "ChildNodes",
            "specifications": Array [
              Object {
                "doNotSort": true,
                "groupByClass": false,
                "groupByLabel": false,
                "relatedClasses": Object {
                  "classNames": Array [
                    "SpatialElement",
                  ],
                  "schemaName": "Biscore",
                },
                "relationships": Object {
                  "classNames": Array [
                    "DistributionSystemGroupsDistributionElements",
                  ],
                  "schemaName": "DistributionSystems",
                },
                "requiredDirection": "Forward",
                "specType": "RelatedInstanceNodes",
              },
            ],
          },
          Object {
            "condition": "ParentNode.IsOfClass(\\"CompositeElement\\", \\"SpatialComposition\\")",
            "ruleType": "ChildNodes",
            "specifications": Array [
              Object {
                "doNotSort": true,
                "groupByClass": false,
                "groupByLabel": false,
                "relatedClasses": Object {
                  "classNames": Array [
                    "CompositeElement",
                  ],
                  "schemaName": "SpatialComposition",
                },
                "relationships": Object {
                  "classNames": Array [
                    "CompositeComposesSubComposites",
                  ],
                  "schemaName": "SpatialComposition",
                },
                "requiredDirection": "Forward",
                "specType": "RelatedInstanceNodes",
              },
            ],
          },
          Object {
            "condition": "ParentNode.IsOfClass(\\"CompositeElement\\", \\"SpatialComposition\\")",
            "ruleType": "ChildNodes",
            "specifications": Array [
              Object {
                "groupByClass": true,
                "groupByLabel": false,
                "relatedClasses": Object {
                  "classNames": Array [
                    "SpatialElement",
                  ],
                  "schemaName": "BisCore",
                },
                "relationships": Object {
                  "classNames": Array [
                    "CompositeOverlapsSpatialElements",
                  ],
                  "schemaName": "SpatialComposition",
                },
                "requiredDirection": "Forward",
                "specType": "RelatedInstanceNodes",
              },
            ],
          },
          Object {
            "condition": "ParentNode.IsOfClass(\\"CompositeElement\\", \\"SpatialComposition\\")",
            "ruleType": "ChildNodes",
            "specifications": Array [
              Object {
                "groupByClass": true,
                "groupByLabel": false,
                "relatedClasses": Object {
                  "classNames": Array [
                    "SpatialElement",
                  ],
                  "schemaName": "BisCore",
                },
                "relationships": Object {
                  "classNames": Array [
                    "SpatialOrganizerReferencesSpatialElements",
                  ],
                  "schemaName": "SpatialComposition",
                },
                "requiredDirection": "Forward",
                "specType": "RelatedInstanceNodes",
              },
            ],
          },
          Object {
            "condition": "ParentNode.IsOfClass(\\"CompositeElement\\", \\"SpatialComposition\\")",
            "ruleType": "ChildNodes",
            "specifications": Array [
              Object {
                "groupByClass": true,
                "groupByLabel": false,
                "relatedClasses": Object {
                  "classNames": Array [
                    "SpatialElement",
                  ],
                  "schemaName": "BisCore",
                },
                "relationships": Object {
                  "classNames": Array [
                    "SpatialOrganizerHoldsSpatialElements",
                  ],
                  "schemaName": "SpatialComposition",
                },
                "requiredDirection": "Forward",
                "specType": "RelatedInstanceNodes",
              },
            ],
          },
          Object {
            "condition": "ContentDisplayType = \\"RelatedElementsRequest\\"",
            "ruleType": "Content",
            "specifications": Array [
              Object {
                "relationshipPaths": Array [
                  Array [
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "CompositeComposesSubComposites",
                        "schemaName": "SpatialComposition",
                      },
                    },
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "CompositeOverlapsSpatialElements",
                        "schemaName": "SpatialComposition",
                      },
                    },
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "ElementOwnsChildElements",
                        "schemaName": "BisCore",
                      },
                    },
                  ],
                ],
                "specType": "ContentRelatedInstances",
              },
              Object {
                "relationshipPaths": Array [
                  Array [
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "CompositeComposesSubComposites",
                        "schemaName": "SpatialComposition",
                      },
                      "targetClass": Object {
                        "className": "ISpatialOrganizer",
                        "schemaName": "SpatialComposition",
                      },
                    },
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "SpatialOrganizerHoldsSpatialElements",
                        "schemaName": "SpatialComposition",
                      },
                    },
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "ElementOwnsChildElements",
                        "schemaName": "BisCore",
                      },
                    },
                  ],
                ],
                "specType": "ContentRelatedInstances",
              },
              Object {
                "relationshipPaths": Array [
                  Array [
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "CompositeComposesSubComposites",
                        "schemaName": "SpatialComposition",
                      },
                    },
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "SpatialOrganizerReferencesSpatialElements",
                        "schemaName": "SpatialComposition",
                      },
                    },
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "ElementOwnsChildElements",
                        "schemaName": "BisCore",
                      },
                    },
                  ],
                ],
                "specType": "ContentRelatedInstances",
              },
              Object {
                "relationshipPaths": Array [
                  Array [
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "DistributionSystemGroupsDistributionElements",
                        "schemaName": "DistributionSystems",
                      },
                    },
                    Object {
                      "count": "*",
                      "direction": "Forward",
                      "relationship": Object {
                        "className": "ElementOwnsChildElements",
                        "schemaName": "BisCore",
                      },
                    },
                  ],
                ],
                "specType": "ContentRelatedInstances",
              },
            ],
          },
        ],
      }
    }
    optionItems={Array []}
    searchTools={true}
    setIsDisplayGuids={[Function]}
    treeName="TreeWithRulesetTest"
    treeNodeIconMapper={
      TreeNodeFunctionIconInfoMapper {
        "_TreeNodeClassFunctionIconInfoMap": Map {},
        "_functionIconInfos": Array [],
        "_globalFunctionIconInfos": Array [],
        "_groupNodeFunctionIconInfos": Array [],
        "_treeDataProvider": Object {
          "dispose": [Function],
          "getFilteredNodePaths": [Function],
          "getNodeKey": [Function],
          "getNodes": [Function],
          "getNodesCount": [Function],
          "imodel": null,
          "loadHierarchy": [Function],
          "rulesetId": "ui-framework/RulesList",
        },
      }
    }
  />
</ContextProvider>
`;
